fgpyo/vcf/builder.py

```diff
@@ -0,0 +1,159 @@
+            sd: optional sequence dictionary
+        """
+        super().__init__(samples=samples, sd=sd)
+        self.header = PysamHeader()
+        for line in VariantBuilder.build_header_string(samples=samples, sd=self.sd).split("\n"):
+            if not line.startswith("##"):
+                break
+            self.header.add_line(line)
+        if samples is not None:
+            self.header.add_samples(samples)
+
+    def close(self) -> None:
+        """Closes this builder"""
+        pass
+
+    def add(
+        self,
+        contig: Optional[str] = None,
+        pos: Optional[int] = None,
+        id: Optional[str] = None,
+        ref: Optional[str] = None,
+        alts: Optional[Iterable[str]] = None,
+        qual: Optional[int] = None,
+        filter: Optional[Iterable[str]] = None,
+        info: Optional[Dict[str, Any]] = None,
+        samples: Optional[Dict[str, Dict[str, Any]]] = None,
+        input_start_index: int = 1,
+    ) -> PysamVariant:
+        """Generates a new variant and adds it to the internal collection.
+
+        Very little validation is done with respect to INFO and FORMAT keys being defined in the
+        header.
+
+        Args:
+            contig: the chromosome name
+            pos: the position
+            id: the variant id
+            ref: the reference allele
+            alts: the list of alternate alleles, None if no alternates
+            qual: the variant quality
+            filter: the list of filters, None if no filters (ex. PASS)
+            info: the dictionary of INFO key-value pairs
+            formats: the dictionary of sample name to FORMAT key-value pairs.
+            input_start_index: the genomic start index for input data (e.g. 0 for 0-based, 1-for
+                               1-based). VCF files are 1-based, but pysam is 0-based.
+        """
+        contig, pos, id, ref, alts, qual = VariantBuilderDefaults.init(
+            self.sd, contig, pos, id, ref, alts, qual
+        )
+
+        if contig not in self.sd:
+            raise ValueError(f"Chromosome `{contig}` not in the sequence dictionary.")
+        if samples is not None and any(sample not in self.samples for sample in samples):
+            raise ValueError(
+                "Extra sample(s) given: "
+                + ", ".join(sample for sample in samples if sample not in self.samples)
+            )
+
+        alleles = (ref,) if alts is None else (ref, *alts)
+
+        if self.samples is None or len(self.samples) == 0:
+            formats = None
+        elif samples is None or len(samples) == 0:
+            formats = [{"GT": "./."} for _ in self.samples]
+        else:
+            format_keys = sorted({key for sample in self.samples for key in samples[sample]})
+            formats = [
+                None
+                if sample not in samples
+                else {key: samples[sample].get(key, ".") for key in format_keys}
+                for sample in self.samples
+            ]
```

Comment:
I reworked some things:
1. I renamed `VariantBuilder.samples` -> `VariantBuilder.sample_ids`
2. I aliased the input `samples` variable to `sample_formats`
3. I renamed the value passed to the `new_record()` function to `record_samples`
4. I was able to substantially simplify the construction of `record_samples`, because pysam actually does most of that stuff for you. The only necessary items was re-ordering by sample_id. It should be much easier to understand now.
(and I added comments explaining why everything was named the way it was)